### DISCUSSION TITLE
how to assign bond orders and keep hydrogens?

### DISCUSSION THREAD

**Discussion by YoavShamir5** [rdkit/rdkit#6911] [Q&A]
Hello, I would appreciate any input on the following - my input is a ligand docking pose in PDB format, which includes hydrogens. In order to generate an SDF version with bond information, I make a template molecule from the correct SMILES pattern. Then, I use AssignBondOrdersFromTemplate, which fails If I do not remove hydrogens from my ligand beforehand. My question is - is there a straightforward way to re-add the exact same hydrogens that were removed, after using AssignBondOrdersFromTemplate? To be clear, I want to add the same hydrogens, with the same connectivity and the same 3D coordinates (AddHs is not sufficient). Thanks!

**Comment by bp-kelley** [comment]
If you want to match explicit hydrogen's, your template MUST have the hydrogens as well.

The docs in AssignBondOrdersFromTemplate are slightly wrong

```
   Note that the template molecule should have no explicit hydrogens
    else the algorithm will fail.
```

should be:

```
   Note that if the pdb molecule has no explicit hydrogens then the template molecule 
   should also have no explicit hydrogens  else the algorithm will fail.
```

Simple example:

```
# make a template
template = Chem.MolFromSmiles("CC=CN")
# add hydrogens (you probably will supply these)
templateh = Chem.AddHs(template)


# read in CCCN with the hydrogens as if it were CC=CN
ps = Chem.SmilesParserParams()
ps.removeHs = False
mol = Chem.MolFromSmiles("C([H])([H])([H])C([H])C([H])N([H])([H])", ps)

# match the template
nmol = AllChem.AssignBondOrdersFromTemplate(templateh, mol)
print(Chem.MolToSmiles(nmol))
```

results in

```
[H]C(=C([H])C([H])([H])[H])N([H])[H]
```

**Reply by YoavShamir5** [reply]
Thank you for the clarification regarding the documentation! My issue persists, because my case is slightly different - I have no explicit hydrogens for the template, but I do have explicit hydrogens for the molecule from the PDB file. Sanitization within AssignBondOrdersFromTemplate fails, unless I set removeHs to True for the molecule. 

mol_16 = Chem.MolFromSmiles("C=CC(Nc1cc(Oc(nc(Nc2ccc(N3CCN(CC3)C)cc2)n4)c5c4[nH]cc5)ccc1)=O")
mol = Chem.MolFromPDBFile("test_mol.pdb", removeHs = False) #see PDB file attached, attached as .txt 
newMol = AllChem.AssignBondOrdersFromTemplate(mol_16,mol)

**AtomValenceException: Explicit valence for atom # 21 N, 5, is greater than permitted**

As I mentioned, removeHs = True solves this issue, but then I lose the explicit hydrogens from the PDB file, along with their coordinates.

[test_mol.txt](https://github.com/rdkit/rdkit/files/13362021/test_mol.txt)
